### PR TITLE
Elo-System

### PR DESCRIPTION
**Pull request by deen13**
Beim Speichern von Spielen wird jedem Team nun ein Elo-Score zugewiesen. Der Elo-Score für neue Spieler ist 1000.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -77,6 +79,8 @@ dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
     implementation "org.jetbrains.kotlin:kotlin-reflect"
     implementation "org.springframework.boot:spring-boot-starter-web"
+    implementation "org.springframework.boot:spring-boot-starter-data-jpa"
+    implementation "com.h2database:h2"
```
Comment (rubengees) [inline build.gradle]:
```suggestion
    runtime "com.h2database:h2"
```

File: src/main/kotlin/de/smartsquare/kickchain/kickway/Blockchain.kt [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
+data class Blockchain(@JsonProperty("chain") val blocks: List<Block>) {
 
     data class Block(@JsonProperty("content") val games: List<Game>?) {
 
         data class Game(val team1: Team, val team2: Team, val score: Score) {
 
-            data class Team(val players: List<String>)
+            data class Team(val players: List<String>) {
+                val first by lazy { players[0] }
+                val second by lazy { players[1] }
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/Blockchain.kt:13]: Aren't there teams with just one player?

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package de.smartsquare.kickchain.kickway.elo
+
+import de.smartsquare.kickchain.kickway.Blockchain
+
+/**
+ * @see https://de.wikipedia.org/wiki/Elo-Zahl (Anpassung der Elo-Zahl)
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: German documentation?

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        one.copy(matches = one.matches + 1, elo = newEloOne),
+        other.copy(matches = other.matches + 1, elo = newEloOther)
+    )
+}
+
+/**
+ * @see https://de.wikipedia.org/wiki/Elo-Zahl (Anpassung der Elo-Zahl)
+ */
+fun Blockchain.Block.Game.matchFactor(firstPlayer: String, secondPlayer: String) =
+    if (this.team1.players.containsAll(listOf(firstPlayer, secondPlayer)) && this.score.goals1 == 10) {
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: https://en.wikipedia.org/wiki/Law_of_Demeter

Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: `when` might look better here.

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloRating.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import javax.persistence.EmbeddedId
+import javax.persistence.Entity
+
+@Entity
+data class EloRating(
+    @EmbeddedId
+    val team: Team,
+    val elo: Double,
+    val matches: Int
+) : Serializable {
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloRating.kt]: Why `Serializable`?

File: src/main/kotlin/de/smartsquare/kickchain/kickway/storing/GameController.kt [hunk 6]
```diff
@@ -24,7 +30,16 @@ class GameController(private val gameRepository: GameRepository) {
             return ResponseEntity.badRequest().body("A team cannot score minus goals")
         }
 
+        val eloOne = eloRatingRepository.findEloRatingByTeamFirstAndTeamSecond(game.team1.first, game.team1.second)
+            .orElse(EloRating(game.team1.first, game.team1.second, elo = 1000.0, matches = 0))
+        val eloTwo = eloRatingRepository.findEloRatingByTeamFirstAndTeamSecond(game.team2.first, game.team2.second)
+            .orElse(EloRating(game.team2.first, game.team2.second, elo = 1000.0, matches = 0))
+        val (newEloOne, newEloTwo) = game.readjust(eloOne, eloTwo)
+        eloRatingRepository.save(newEloOne)
+        eloRatingRepository.save(newEloTwo)
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/storing/GameController.kt]: This should happen in a service with `@Transactional` to guarantee consistency.

File: src/test/groovy/de/smartsquare/kickchain/kickway/storing/GameControllerSpecfication.groovy [hunk 7]
```diff
@@ -107,4 +112,83 @@ class GameControllerSpecfication extends Specification {
         then:
         verify(gameRepository, times(1)).save(any(Blockchain.Block.Game.class))
     }
+
+    def 'initial elo rating is assigned to the winner team'() {
+        setup:
+        eloRatingRepository.deleteAll()
```
Comment (rubengees) [inline src/test/groovy/de/smartsquare/kickchain/kickway/storing/GameControllerSpecfication.groovy]: Maybe do this globally in `tearDown`?

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    else -> 20
+}
+
+/**
+ * @see https://de.wikipedia.org/wiki/Elo-Zahl (Erwartungswert)
+ */
+infix fun EloRating.odds(opposite: EloRating) =
+    (1 / (1 + Math.pow(10.0, (opposite.elo - this.elo) / 400))).roundToThreeDecimals()
+
+private fun Double.roundToThreeDecimals() = Math.round(this * 1000.0) / 1000.0
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: `BigDecimal` is better for this: https://www.tutorialspoint.com/java/math/bigdecimal_round.htm

**Review by rubengees — DISMISSED** [review]
Cool!

File: src/test/groovy/de/smartsquare/kickchain/kickway/storing/GameControllerSpecfication.groovy (hunk 7 above)
Comment (deen13) [reply src/test/groovy/de/smartsquare/kickchain/kickway/storing/GameControllerSpecfication.groovy]: I replaced the setup hooks by @DirtiesContext

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt (hunk 8 above)
Comment (deen13) [reply src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: I am using BigDecimal now but what did we gained? 🤔

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt (hunk 4 above)
Comment (deen13) [reply src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: Thanks! I came up with this solution:

```kotlin
this.team1.consistsOf(firstPlayer and secondPlayer) && this.score.goals1 == 10 -> 1
```

File: src/main/kotlin/de/smartsquare/kickchain/kickway/Blockchain.kt (hunk 2 above)
Comment (deen13) [reply src/main/kotlin/de/smartsquare/kickchain/kickway/Blockchain.kt:13]: darn

File: src/test/groovy/de/smartsquare/kickchain/kickway/storing/GameControllerSpecfication.groovy (hunk 7 above)
Comment (rubengees) [reply src/test/groovy/de/smartsquare/kickchain/kickway/storing/GameControllerSpecfication.groovy]: Hmm, that makes tests significantly slower. Does above not work?

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt (hunk 8 above)
Comment (rubengees) [reply src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: `doubles` and `floats` round incorrectly in corner cases, here is more info: https://stackoverflow.com/questions/2808535/round-a-double-to-2-decimal-places

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt (hunk 4 above)
Comment (rubengees) [reply src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: Like it :slightly_smiling_face:

File: src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    when {
+        this.team1.consistsOf(firstPlayer and secondPlayer) && this.score.goals1 == 10 -> 1
+        this.team2.consistsOf(firstPlayer and secondPlayer) && this.score.goals2 == 10 -> 1
+        else -> 0
+    }
+
+/**
+ * This is just a fluent wrapper for the to infix function.
+ */
+private infix fun String.and(other: String) = this to other
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/elo/EloCalculator.kt]: What is the point of this? Just use `to`?

**Pull request merged**